Use BibTeX citations in Doxygen

jngrad opened the pull request:
Follow-up to the Summer School developers meeting

Description of changes:
- replace all hardcoded literature references in the core by BibTeX references (817e59b00be)
- backport improvements from 4.1.1: grammar and spell check, copyright years
- remove unit tests from the doxygen documentation (too much clutter)
- cleanup doxygen documentation

jngrad committed:
Cleanup doxygen

Move docstrings above the corresponding functions, remove duplicate
docstrings, transform regular comments into docstrings if relevant,
fix typos, remove outdated text, typeset math formula.

jngrad committed:
Enable citations in Doxygen

jngrad committed:
Use DOI as link

The bibliographic styles available to Doxygen do not process the
DOI field of BibTeX files. The DOI field is now used to create the
URL field programmatically.

jngrad committed:
Replace hardcoded literature references by BibTeX

The citations were obtained from icp.bib.

jngrad committed:
Run Doxygen on CUDA 10

Doxygen now requires a BibTeX installation.

jngrad committed:
Formatting

jngrad committed:
Update copyright year

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+  volume                   = {271},
+  doi                      = {10.1017/S0022112094001771},
+}
+@article{sonnenschein85a,
```
Please use consistent capitalization, i.e., `@Article`.

jngrad replied:
done

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+year = {1985},
+doi = {10.1016/0021-9991(85)90151-2},
+}
+@article{neumann85b,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+  pages     = {327--348},
+  doi       = {10.1016/s0010-4655(02)00586-6},
+}
+@article{marsili10a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+  volume                   = {65},
+  doi                      = {10.1103/PhysRevE.65.046308},
+}
+@incollection{burtscher11a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+  volume                   = {104},
+  doi                      = {10.1023/A:1010414013942},
+}
+@article{dupin08a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+year  = {2008},
+doi = {10.1080/10618560802238242},
+}
+@inproceedings{zhang01b,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+isbn={978-3-8348-2376-2},
+doi={10.1007/978-3-8348-2376-2},
+}
+@article{gompper96a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+year={1996},
+doi={10.1051/jp1:1996246},
+}
+@article{moss96a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+pages={935-938},
+doi={10.1109/ICIP.2001.958278},
+}
+@book{kruger12a,
```
Capitalization.

pkreissl commented on doc/doxygen/bibliography.bib:
```diff
@@ -0,0 +1,323 @@
+@Article{usta05a,
```
Actually, I think it would make a lot of sense to have the entries in this file in alphanumerical order.

jngrad replied:
Done.

pkreissl commented on src/core/electrostatics_magnetostatics/icc.hpp:
```diff
@@ -24,16 +24,13 @@
- *  C. Holm: An iterative, fast, linear-scaling method for computing induced
- *  charges on arbitrary dielectric boundaries, J. Chem. Phys. 2010, 132,
- *  p. 154112, doi:10.1063/1.3376011
+ *  self-consistently using an iterative scheme.  It can at presently -
```
Correct wrong grammar/typo: "It can at presently [...]"

jngrad replied:
Thanks!

pkreissl commented on src/core/electrostatics_magnetostatics/icc.hpp:
```diff
@@ -24,16 +24,13 @@
+ *  details see: @cite tyagi10a
  *
- *  To set up ICCP3M first the dielectric boundary has to be modelled
+ *  To set up ICCP3M, first the dielectric boundary has to be modelled
```
AE: modeled

pkreissl commented on src/core/electrostatics_magnetostatics/icc.hpp:
```diff
@@ -58,21 +55,21 @@
-  double relax = 0.7; /* relaxation parameter for iterative */
-  int citeration = 0; /* current number of iterations*/
-  int first_id = 0;
+      ein; /**< Array of dielectric constants at each surface element */
```
That line should probably be joined to the line above -- if then the comment becomes too long, the description can be line-broken.

pkreissl commented on src/core/electrostatics_magnetostatics/mmm2d.hpp:
```diff
@@ -19,14 +19,14 @@
+ *  Implementation of the MMM2D method for the calculation
+ *  of the electrostatic interaction for two dimensionally periodic systems.
+ *  For details on the method see MMM general. The MMM2D method works only with
+ *  the layered or nsquared  "cell system".  The tuning is not automated, since
```
Double spaces before and after "cell system".

pkreissl commented on src/core/grid_based_algorithms/lb.cpp:
```diff
@@ -1311,8 +1312,8 @@ Utils::Vector6d lb_calc_stress(std::array<double, 19> const &modes,
   // correspond to those used by U. Schiller. In terms of populations this
-  // expression then corresponds exactly to those in Eqs. 116 - 121 in the
-  // Duenweg and Ladd paper, when these are written out in populations.
+  // expression then corresponds exactly to those in Eqs. 116 - 121 in
```
116-121 (no spaces), your call.

pkreissl replied:
Or rather, to be consistent with other equations cited: (116)-(121)

pkreissl commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -275,100 +275,108 @@ __device__ __inline__ float calc_mode_x_from_n(LB_nodes_gpu n_a,
+   *
+   *  The basis vectors (modes) are constructed as follows (eq. 111):
+   *  \f[m_k = \sum_{i} e_{ki} n_{i}\f] where the \f$e_{ki}\f$ form a
+   *  linear transformation (matrix) that is given by (modified from Table 1):
```
**t**able

pkreissl commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -275,100 +275,108 @@ __device__ __inline__ float calc_mode_x_from_n(LB_nodes_gpu n_a,
+   *                  1/36, 1/36, 1/36, 1/36,
+   *                  1/36, 1/36, 1/36 )\f]
+   *
+   *  Which makes the transformation satisfy the following
```
**w**hich

pkreissl commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -1143,8 +1150,8 @@ calc_values_in_LB_units(LB_nodes_gpu n_a, Utils::Array<float, 19> &mode,
     // correspond to those used by U. Schiller. In terms of populations this
-    // expression then corresponds exactly to those in Eqs. 116 - 121 in the
-    // Duenweg and Ladd paper, when these are written out in populations.
+    // expression then corresponds exactly to those in Eqs. 116 - 121 in
```
(116)-(121)

jngrad committed:
Document and improve readability of sed expression

pkreissl commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -275,100 +275,108 @@ __device__ __inline__ float calc_mode_x_from_n(LB_nodes_gpu n_a,
+   *   c_{18} &= ( 0,-1, 1)
+   *  \f}
+   *
+   *  The basis vectors (modes) are constructed as follows (eq. 111):
```
`(eq. (111) @cite duenweg09a)`

jngrad replied:
I've now put `@cite duenweg09a` in the first line, since the same paper is used for all equations in this doxygen block.

pkreissl commented on src/core/grid_based_algorithms/lbgpu_cuda.cu:
```diff
@@ -275,100 +275,108 @@ __device__ __inline__ float calc_mode_x_from_n(LB_nodes_gpu n_a,
+   *                  1/36, 1/36, 1/36 )\f]
+   *
+   *  Which makes the transformation satisfy the following
+   *  orthogonality condition (eq. 109):
```
`(eq. (109) @cite duenweg09a)`

pkreissl commented on src/core/immersed_boundary/ImmersedBoundaries.cpp:
```diff
@@ -312,8 +313,8 @@ void ImmersedBoundaries::calc_volume_force() {
 
-            // This is Dupin 2008. I guess the result will be very similar as
-            // the code above
+            // This is eq. (9) in @cite dupin08a. I guess the result will be
```
Is this seriously a in-code documentation comment: "I guess the result will be very similar as the code above" ??? + missing punctuation.

jngrad replied:
This file has 20% coverage. IBM is essentially unmaintained. It has been a candidate for downgrade or removal for more than a year (#2071, #2305).

pkreissl commented on src/core/immersed_boundary/ibm_triel.hpp:
```diff
@@ -23,15 +23,19 @@
-// This function is used to set the parameters
-// Also calculates and stores the reference state
+/** Set the IBM Triel parameters.
+ *  Also calculate and store the reference state
```
Missing full stop.

pkreissl commented on src/core/rattle.hpp:
```diff
@@ -22,11 +22,9 @@
- * RATTLE Algorithm (Rattle: A "Velocity" Version of the
- * Shake Algorithm for Molecular Dynamics Calculations, H.C Andersen, J Comp
- * Phys, 52, 24-34, 1983)
+ *  RATTLE Algorithm (@cite andersen83a)
```
**a**lgorithm

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers.cpp:
```diff
@@ -148,12 +142,7 @@ void IBM_UpdateParticlePositions(ParticleRange particles) {
-CPU
-**************/
-
+/** Put the momentum of a given particle into the LB fluid */
```
Missing full stop.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers.cpp:
```diff
@@ -188,23 +177,20 @@ void CoupleIBMParticleToFluid(Particle *p) {
-*******************/
-
+/** Calculate the LB fluid velocity at a particle position.
+ *  Very similar to the velocity interpolation done in standard Espresso,
```
write "Espresso" the proper way...

jngrad replied:
Now fixed everywhere.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers.cpp:
```diff
@@ -188,23 +177,20 @@ void CoupleIBMParticleToFluid(Particle *p) {
+ *  Very similar to the velocity interpolation done in standard Espresso,
+ *  except that we add the f/2 contribution, cf. @cite guo02a.
+ *  The fluid velocity is obtained by linear interpolation,
+ *  cf. Eq. (11) in @cite ahlrichs99a.
```
This is maybe going too far, however there is quite some inconsistency concerning capitalization of mid-sentence abbreviations "Eq.", "Eqs." vs. "eq.", "eqs.".

jngrad replied:
I'm now only using a capital `Eq.` at the start of a sentence.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers.cpp:
```diff
@@ -313,12 +294,9 @@ bool IsHalo(const int indexCheck) {
-structure
-*****************/
-
+/** Get particle velocities from LB and set the velocity field in the particles
```
Missing full stop.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers_cuda.cu:
```diff
@@ -68,16 +68,12 @@ extern LB_node_force_density_gpu node_f;
 // ** These variables are static in lbgpu_cuda.cu, so we need to duplicate them
-// here They are initialized in ForcesIntoFluid The pointers are on the host,
+// here. They are initialized in ForcesIntoFluid. The pointers are on the host,
 // but point into device memory
```
Missing full stop.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers_cuda.cu:
```diff
@@ -68,16 +68,12 @@ extern LB_node_force_density_gpu node_f;
-Calls a kernel to reset the forces on the LB nodes to the external force
-*****************/
-
+/** Call a kernel to reset the forces on the LB nodes to the external force */
```
Missing full stop.

pkreissl commented on src/core/virtual_sites/lb_inertialess_tracers_cuda.cu:
```diff
@@ -93,13 +89,11 @@ void IBM_ResetLBForces_GPU() {
+ *  Called from @ref integrate_vv.
+ *  This must be the first CUDA-IBM function to be called because it also does
+ *  some initialization
+ */
```
Full stop, again.

jngrad committed:
Fix typos and capitalization

jngrad committed:
Exclude unit tests from the documentation

They do not provide any useful information and pollute the
documentation with irrelevant global variables and mocked
structs/classes that have the same name as the corresponding
real structs/classes, making it harder to find the relevant
documentation page.

jngrad committed:
Sort BibTeX records alphabetically

pkreissl commented on src/core/electrostatics_magnetostatics/p3m-dipolar.hpp:
```diff
@@ -163,7 +161,7 @@ void dp3m_deactivate();
+ *  for the dipolar P3M method in the paper of @cite cerda08d in
  *  order to obtain the rms error in the force for a system of N randomly
  *  distributed particles in a cubic box.
  *  For the real space error the estimate of Kolafa/Perram is used.
```
Just wondering, is there a citation for that one?

jngrad replied:
I love references with the author name but no year and no journal...

jngrad committed:
Cleanup doxygen

jngrad committed:
Add more BibTeX records

jngrad commented:
@pkreissl I think I got all of them this time. I couldn't deduce the reference for the following 4 citations that have no year information: Krueger and Fedosov, Yeh and Klapp, Schiller, Moshier.

pkreissl commented on src/core/electrostatics_magnetostatics/mmm-common.hpp:
```diff
@@ -19,8 +19,14 @@
-    modified polygamma functions. See Arnold,Holm 2002
-*/
+ *  Common parts of the MMM family of methods for the electrostatic
+ *  interaction, MMM1D, MMM2D and ELC.  This file contains the code for the
```
Double space after "ELC."

jngrad replied:
I think the double whitespace after a dot was a recommendation in the early days of UNIX. But in ESPResSo it is rarely used, except in license headers. I've removed them outside licenses.

pkreissl reviewed (approved):
I found another double space in the code documentation, your call.
Otherwise you're good to go.

jngrad committed:
Cleanup doxygen

Remove outdated information, fix typos, remove double whitespaces,
fix broken links, fix doxygen style.

jngrad committed:
Re-shuffle doxygen blocks

jngrad commented:
bors r=fweik
